### PR TITLE
[MM-20804] Add "Unarchive Channel" option to the channel info screen

### PR DESCRIPTION
**Pull request by allenlai18**
#### Summary

When restoring channel, publish system message to channel via websocket

#### Related PR on other repositories
* Mobile PR:  https://github.com/mattermost/mattermost-mobile/pull/3663
* Redux PR:  https://github.com/mattermost/mattermost-redux/pull/1004
* Webapp PR: https://github.com/mattermost/mattermost-webapp/pull/4458
* Server PR: https://github.com/mattermost/mattermost-server/pull/13349

#### Ticket Link
JIRA ticket: https://mattermost.atlassian.net/browse/MM-20804
GH issue: https://github.com/mattermost/mattermost-server/issues/13307

### CODE REVIEW

File: app/channel.go [hunk 1]
```diff
@@ -596,10 +596,90 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+
+	now := model.GetMillis()
+	for _, hook := range incomingHooks {
+		if err := a.Srv.Store.Webhook().DeleteIncoming(hook.Id, now); err != nil {
+			mlog.Error("Encountered error deleting incoming webhook", mlog.String("hook_id", hook.Id), mlog.Err(err))
+		}
+		a.InvalidateCacheForWebhook(hook.Id)
+	}
+	for _, hook := range outgoingHooks {
+		if err := a.Srv.Store.Webhook().DeleteOutgoing(hook.Id, now); err != nil {
```
Comment (hanzei) [inline app/channel.go]: Can you elaborate why we need to delete the webhooks again? There are already deleted when the channel is archived.

File: i18n/en.json [hunk 2]
```diff
@@ -47,10 +47,6 @@
     "id": "September",
     "translation": "September"
   },
-  {
-    "id": "actiance.export.marshalToXml.appError",
-    "translation": "Unable to convert export to XML."
-  },
```
Comment (hanzei) [inline i18n/en.json]: There are a lot of unrelated changes in the file

File: app/channel.go [hunk 3]
```diff
@@ -596,10 +596,90 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
 		return nil, err
 	}
+
+	if user != nil {
+		T := utils.GetUserTranslations(user.Locale)
+
+		post := &model.Post{
+			ChannelId: channel.Id,
+			Message:   fmt.Sprintf(T("api.channel.undelete_channel.unarchived"), user.Username),
+			Type:      model.POST_CHANNEL_UNDELETED,
```
Comment (hanzei) [inline app/channel.go]: I'm wondering, if we really need another post type for this?

File: i18n/en.json (hunk 2 above)
Comment (allenlai18) [reply i18n/en.json]: yeah. im not sure. i ran ```make i18n-extract```

File: app/channel.go (hunk 1 above)
Comment (allenlai18) [reply app/channel.go]: I am still learning about webhooks. 😅. you are right. We do not need to delete them.  Since we are unarchiving them. do we need to create incoming and outgoing webhooks for the channel?

File: app/channel.go [hunk 4]
```diff
@@ -596,10 +596,48 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
 	return nil
 }
 
-func (a *App) RestoreChannel(channel *model.Channel) (*model.Channel, *model.AppError) {
+func (a *App) RestoreChannel(channel *model.Channel, userId string) (*model.Channel, *model.AppError) {
+	var user *model.User
+	if userId != "" {
```
Comment (hanzei) [inline app/channel.go]: I think we are fine to require `userId` to be ` != ""` and drop this check.

File: app/channel.go [hunk 5]
```diff
@@ -596,10 +596,48 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+		var err *model.AppError
+		user, err = a.Srv.Store.User().Get(userId)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if channel.DeleteAt == 0 {
+		err := model.NewAppError("undeleteChannel", "api.channel.undelete_channel.undeleted.app_error", nil, "", http.StatusBadRequest)
+		return nil, err
```
Comment (hanzei) [inline app/channel.go]: You can combine these two lines

File: app/channel.go [hunk 6]
```diff
@@ -596,10 +596,48 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+	var user *model.User
+	if userId != "" {
+		var err *model.AppError
+		user, err = a.Srv.Store.User().Get(userId)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if channel.DeleteAt == 0 {
```
Comment (hanzei) [inline app/channel.go]: Nit: You might want to move this check before the code that gets the users, because it doesn't require a DB lookup.

File: app/channel.go (hunk 3 above)
Comment (hanzei) [reply app/channel.go]: There is `model.POST_SYSTEM_GENERIC` for generic system posts. If do specific rendering on web or mobile is needed, we can use this.

File: app/channel.go [hunk 7]
```diff
@@ -596,10 +596,48 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+	if user != nil {
+		T := utils.GetUserTranslations(user.Locale)
+
+		post := &model.Post{
+			ChannelId: channel.Id,
+			Message:   fmt.Sprintf(T("api.channel.undelete_channel.unarchived"), user.Username),
+			Type:      model.POST_CHANNEL_UNDELETED,
+			UserId:    userId,
+			Props: model.StringInterface{
+				"username": user.Username,
```
Comment (hanzei) [inline app/channel.go]: Why do we need to include the `username`?

File: app/channel.go [hunk 8]
```diff
@@ -596,10 +596,48 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+			},
+		}
+
+		if _, err := a.CreatePost(post, channel, false); err != nil {
+			mlog.Error("Failed to post unarchive message", mlog.Err(err))
+		}
+	}
+
 	if err := a.Srv.Store.Channel().Restore(channel.Id, model.GetMillis()); err != nil {
```
Comment (hanzei) [inline app/channel.go]: I think we want to first restore the channel and then add the post

File: i18n/en.json [hunk 9]
```diff
@@ -263,6 +263,14 @@
     "id": "api.channel.delete_channel.type.invalid",
     "translation": "Unable to delete direct or group message channels"
   },
+  {
+    "id": "api.channel.undelete_channel.unarchived",
+    "translation": "%v unarchived the channel."
+  },
+  {
+    "id": "api.channel.undelete_channel.undeleted.app_error",
+    "translation": "The channel has been unarchived"
```
Comment (hanzei) [inline i18n/en.json]: That seams like a misleading error message. What about
```suggestion
    "translation": "Unable to unarchived a not archived channel"
```

File: app/channel.go (hunk 1 above)
Comment (hanzei) [reply app/channel.go]: Good question. @wiersgallak should existing webhooks be restored?

File: app/channel.go (hunk 4 above)
Comment (allenlai18) [reply app/channel.go]: not sure what you mean. but are you saying we should just drop this if statement?

File: app/channel.go (hunk 6 above)
Comment (allenlai18) [reply app/channel.go]: right

File: app/channel.go (hunk 5 above)
Comment (allenlai18) [reply app/channel.go]: right

File: app/channel.go (hunk 7 above)
Comment (allenlai18) [reply app/channel.go]: i think we need it because the post says "{username} unarchived the channel"?

File: app/channel.go (hunk 3 above)
Comment (allenlai18) [reply app/channel.go]: what do you mean by specific rendering?

File: app/channel.go (hunk 4 above)
Comment (hanzei) [reply app/channel.go]: Yes

File: app/channel.go (hunk 7 above)
Comment (hanzei) [reply app/channel.go]: You are building the message in https://github.com/mattermost/mattermost-server/pull/13349/files#diff-243b54bc23ef263e3d67c7f01d3972a0R621 with the user name. Why do you need it in `Props`?

Comment (allenlai18) [reply app/channel.go]: youre right. we are grabbing it from the user object

File: app/channel.go (hunk 3 above)
Comment (hanzei) [reply app/channel.go]: The webapp and mobile support specific rendering of post types. E.g. some post type could be displayed with blue text.

File: app/channel.go [hunk 10]
```diff
@@ -596,10 +596,42 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
 		return nil, err
 	}
+	a.InvalidateCacheForChannel(channel)
+
+	var user *model.User
+	var err *model.AppError
+	user, err = a.Srv.Store.User().Get(userId)
```
Comment (hanzei) [inline app/channel.go]: No need to pre declare the variables. This can just be
```suggestion
	user, err := a.Srv.Store.User().Get(userId)
```

File: app/channel.go [hunk 11]
```diff
@@ -596,10 +596,42 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+		}
+
+		if _, err := a.CreatePost(post, channel, false); err != nil {
+			mlog.Error("Failed to post unarchive message", mlog.Err(err))
+		}
+	}
+
+	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_CHANNEL_UNDELETED, channel.TeamId, "", "", nil)
+	message.Add("channel_id", channel.Id)
+	a.Publish(message)
```
Comment (hanzei) [inline app/channel.go]: It's a minor thing, but to be consistent with the rest of the code base these three lines should be after https://github.com/mattermost/mattermost-server/pull/13349/files#diff-243b54bc23ef263e3d67c7f01d3972a0R607

File: app/channel.go [hunk 12]
```diff
@@ -596,10 +596,42 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+	if err != nil {
+		return nil, err
+	}
+
+	if user != nil {
+		T := utils.GetUserTranslations(user.Locale)
+
+		post := &model.Post{
+			ChannelId: channel.Id,
+			Message:   fmt.Sprintf(T("api.channel.undelete_channel.unarchived"), user.Username),
```
Comment (hanzei) [inline app/channel.go]: The cleaner way would be to use the template feature from goi18n and not fmt. Something like
```suggestion
			Message:   T("api.channel.undelete_channel.unarchived", map[string]interface{}{"Username": user.Username}),
```

And change the message in `i18n/en.json` to `{{ .Username }}` unarchived the channel.`

File: app/channel.go (hunk 3 above)
Comment (allenlai18) [reply app/channel.go]: ah. yes we are rendering this system message differently in mobile

File: app/channel.go (hunk 11 above)
Comment (allenlai18) [reply app/channel.go]: didnt realize all the other methods publish this message right after invalidating cache

File: i18n/en.json [hunk 13]
```diff
@@ -263,6 +263,14 @@
     "id": "api.channel.delete_channel.type.invalid",
     "translation": "Unable to delete direct or group message channels"
   },
+  {
+    "id": "api.channel.undelete_channel.unarchived",
```
Comment (hanzei) [inline i18n/en.json]: The translations are ordert by name so this has to be after `api.channel.rename_channel.cant_rename_group_messages.app_error`.

Comment (allenlai18) [reply i18n/en.json]: done

File: i18n/en.json [hunk 14]
```diff
@@ -367,6 +371,10 @@
     "id": "api.channel.rename_channel.cant_rename_group_messages.app_error",
     "translation": "You cannot rename a group message channel"
   },
+  {
+    "id": "api.channel.undelete_channel.unarchived",
```
Comment (hanzei) [inline i18n/en.json]: You forgot to also reorder this one. see https://build.mattermost.com/blue/organizations/jenkins/mp%2Fmattermost-server-pr-new/detail/PR-13349/14/pipeline

File: app/channel.go (hunk 3 above)
Comment (hanzei) [reply app/channel.go]: Just want to double check why you need to render this differently on mobile? Why isn't the "normal" system post fine?

File: i18n/en.json (hunk 14 above)
Comment (allenlai18) [reply i18n/en.json]: 😭 . done.

File: app/channel.go (hunk 3 above)
Comment (allenlai18) [reply app/channel.go]: It is actually rendered basically the same. the archive channel is doing it the same way and is basically the same functionality. so i was kind of going for consistency.  I also see the other post types like HEADER_CHANGE is rendering the same.  🤔

Comment (allenlai18) [reply app/channel.go]: Do we need it anytime we are rendering a Username?  I think its that.

**Comment by hanzei** [comment]
It would be nice to have this covered by a test, but I'M not blocking on this

File: app/channel.go (hunk 1 above)
Comment (wiersgallak) [reply app/channel.go]: @jasonblais or @aaronrothschild Can one of you weigh in on this question, since it's more in your area of ownership?

Comment (jasonblais) [reply app/channel.go]: For the purposes of this PR, I would keep the behaviour consistent when you restore a channel [via the API](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D~1restore%2Fpost) or [via the CLI](https://docs.mattermost.com/administration/command-line-tools.html#mattermost-channel-restore).

So for whatever we do there, we should do the same here, and document it.

Comment (allenlai18) [reply app/channel.go]: the API and CLI documentation doesnt say anything about webhooks.

File: app/channel.go [hunk 15]
```diff
@@ -596,10 +596,40 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
 	return nil
 }
 
-func (a *App) RestoreChannel(channel *model.Channel) (*model.Channel, *model.AppError) {
+func (a *App) RestoreChannel(channel *model.Channel, userId string) (*model.Channel, *model.AppError) {
```
Comment (saturninoabril) [inline app/channel.go]: I agree with @hanzei, please add unit test for this.

File: app/channel.go [hunk 16]
```diff
@@ -596,10 +596,40 @@ func (a *App) postChannelPrivacyMessage(user *model.User, channel *model.Channel
...
+	if channel.DeleteAt == 0 {
+		return nil, model.NewAppError("undeleteChannel", "api.channel.undelete_channel.undeleted.app_error", nil, "", http.StatusBadRequest)
```
Comment (saturninoabril) [inline app/channel.go]: I feel like `RESTORE*/restore*` sounds better and consistent than `UNDELETE*/undelete*`.

File: i18n/en.json [hunk 17]
```diff
@@ -367,6 +367,14 @@
...
+    "translation": "{{.Username}} unarchived the channel."
+  },
+  {
+    "id": "api.channel.undelete_channel.undeleted.app_error",
+    "translation": "Unable to unarchived a not archived channel"
```
Comment (saturninoabril) [inline i18n/en.json]: I'll defer to PM on correct text messages introduced here.

File: app/channel.go (hunk 1 above)
Comment (jasonblais) [reply app/channel.go]: @allenlai18 Sorry for the late reply. You're correct, the documentation does not. Would you be open to trying what happens when you archive and restore a channel containing a webhook via the API (or CLI, whichever is easier for you)? We can then use the same approach here and document it.

File: i18n/en.json [hunk 18]
```diff
@@ -367,6 +367,14 @@
     "id": "api.channel.rename_channel.cant_rename_group_messages.app_error",
     "translation": "You cannot rename a group message channel"
   },
+  {
+    "id": "api.channel.undelete_channel.unarchived",
```
Comment (jasonblais) [inline i18n/en.json]:
```suggestion
    "id": "api.channel.restore_channel.unarchived",
```

Comment (jasonblais) [reply i18n/en.json]: Proposing `restore_channel` to be consistent with the RestoreChannel API method

File: i18n/en.json (hunk 17 above)
Comment (jasonblais) [reply i18n/en.json]: Thank you Saturn for the ping, I missed this during initial review.

@allenlai18 confirming if this error message only occurs when you are trying to restore/unarchive a channel that wasn't previously archived? Or are there any other cases where this error may be shown?

File: app/channel.go (hunk 16 above)
Comment (allenlai18) [reply app/channel.go]: we are using 'Unarchive' with the User interface. 'Undelete' on the client applications. and 'Restore' on the API/server level.   I changed it to Restore

File: i18n/en.json (hunk 17 above)
Comment (allenlai18) [reply i18n/en.json]: Yes. that is the case when this error happens

File: i18n/en.json (hunk 18 above)
Comment (allenlai18) [reply i18n/en.json]: done

File: i18n/en.json [hunk 19]
```diff
@@ -367,6 +367,14 @@
     "id": "api.channel.rename_channel.cant_rename_group_messages.app_error",
     "translation": "You cannot rename a group message channel"
   },
+  {
+    "id": "api.channel.restore_channel.restored.app_error",
+    "translation": "Unable to unarchived a not archived channel"
```
Comment (jasonblais) [inline i18n/en.json]:
```suggestion
    "translation": "Unable to unarchive channel. The channel is not archived."
```

Comment (jasonblais) [reply i18n/en.json]: ^proposed update. Definitely open for feedback, I'm only 1/5 on this particular wording. @wiersgallak wondering if you had any feedback to share,

File: app/channel.go (hunk 15 above)
Comment (allenlai18) [reply app/channel.go]: @saturninoabril  The functionality I am adding is validating and sending a NewWebSocketEvent.   I should add unit test to test this or the actual RestoreChannel functionality?

File: app/channel.go (hunk 1 above)
Comment (allenlai18) [reply app/channel.go]: @jasonblais @hanzei @saturninoabril  when you archive a channel. a message comes up on channel for "channel is archived" and the webhooks are deleted.

when you unarchive a channel. for web the a message 'channel is unarchive' does not show up in realtime but you need to refresh. but for mobile it works. I am not too sure why, i think it has something to do with webhooks. I was hoping to get some help with that.

File: i18n/en.json (hunk 19 above)
Comment (allenlai18) [reply i18n/en.json]: i like this better than the one before. went with this change

File: app/channel.go (hunk 1 above)
Comment (saturninoabril) [reply app/channel.go]: >  i think it has something to do with webhooks
It's with the websocket event/message.  You may revisit how it's handled on webapp.

File: app/channel.go (hunk 15 above)
Comment (saturninoabril) [reply app/channel.go]: Add unit test to RestoreChannel since to test yet on it.  Would be awesome if you could include check to websocket event.

File: model/post.go [hunk 20]
```diff
@@ -38,6 +38,7 @@ const (
 	POST_CONVERT_CHANNEL        = "system_convert_channel"
 	POST_PURPOSE_CHANGE         = "system_purpose_change"
 	POST_CHANNEL_DELETED        = "system_channel_deleted"
+	POST_CHANNEL_UNDELETED      = "system_channel_undeleted"
```
Comment (saturninoabril) [inline model/post.go]: Please change to
```suggestion
	POST_CHANNEL_RESTORED      = "system_channel_restored"
```

File: model/websocket_message.go [hunk 21]
```diff
@@ -18,6 +18,7 @@ const (
 	WEBSOCKET_EVENT_CHANNEL_CONVERTED       = "channel_converted"
 	WEBSOCKET_EVENT_CHANNEL_CREATED         = "channel_created"
 	WEBSOCKET_EVENT_CHANNEL_DELETED         = "channel_deleted"
+	WEBSOCKET_EVENT_CHANNEL_UNDELETED       = "channel_undeleted"
```
Comment (saturninoabril) [inline model/websocket_message.go]:
```suggestion
	WEBSOCKET_EVENT_CHANNEL_RESTORED       = "channel_restored"
```

File: app/channel.go (hunk 3 above)
Comment (hanzei) [reply app/channel.go]: @saturninoabril Can you help with this discussion as I don't know enough about the webapp and mobile code.

File: app/channel.go (hunk 15 above)
Comment (allenlai18) [reply app/channel.go]: wrote a simple unit test.  i am not very experienced in writing unit test and not sure the best way to do it and what cases to test.  If i can get more guidance and help. that would be great.  Would this be better for a separate PR for the amount of work of writing good unit tests? I am also confused by the channel_test.go in api4 and app folder. there seems to be a lot of overlap

**Comment by jasonblais** [comment]
@jespino / @saturninoabril friendly reminder to help with dev/QA reviews when you have time :)

File: app/channel_test.go [hunk 22]
```diff
@@ -1149,3 +1149,26 @@ func TestMarkChannelAsUnreadFromPost(t *testing.T) {
...
+	th.LoginTeamAdmin()
+
+	_, resp = Client.RestoreChannel(publicChannel1.Id)
+	CheckOKStatus(t, resp)
+
+	_, resp = Client.RestoreChannel(privateChannel1.Id)
+	CheckOKStatus(t, resp)
+
+	ch, err := th.App.GetChannel(publicChannel1.Id)
+	require.True(t, err != nil || ch.DeleteAt == 0, "should have restored channel with variable DeleteAt = 0.")
```
Comment (saturninoabril) [inline app/channel_test.go]: Please the same check for `privateChannel1`

**Comment by jasonblais** [comment]
@jespino Friendly reminder to help with reviews when back from the holidays, thanks!

**Comment by hanzei** [comment]
@jespino Could you please confirm that a custom post type is needed? See https://github.com/mattermost/mattermost-server/pull/13349/files/a196655c46c8bb783b7366b4dd7290d3109fd5b0#r355062159

**Comment by jespino** [comment]
@hanzei let me check

File: app/channel.go (hunk 3 above)
Comment (jespino) [reply app/channel.go]: Most of the types are not strictly needed, I think is correct to keep it consistent with other types, like `POST_CHANNEL_DELETED`, the `POST_SYSTEM_GENERIC` shouldn't be an excuse to not properly define the correct types for the posts.

**Comment by jespino** [comment]
Merged! thanks @allenlai18! :dancing_women:
